### PR TITLE
[APM-CI][Integration Tests] Support dotnet agent

### PR DESCRIPTION
**Pull request by v1v**
## Highlights
- Use `opbeans-dotnet` app
- Use test app from https://github.com/elastic/apm-integration-testing/pull/425
- Support release and build options, aka use the release version of the dot-net agent or build from the source code.
- Add steps in the Jenkinsfile/Matrix one.
- Cache docker images.
- Update Docs.

## Test Results
- Some documents are lost when running the concurrent ITs (in the CI and locally)

```
tests/agent/test_dotnet.py::test_dotnet_error PASSED                     [ 25%]
tests/agent/test_dotnet.py::test_concurrent_req_dotnet FAILED            [ 50%]
tests/agent/test_dotnet.py::test_concurrent_req_dotnet_foobar FAILED     [ 75%]
tests/agent/test_dotnet.py::test_req_dotnet FAILED                       [100%]

# 1st test execution
expected 1375, got 1319
expected 1375, got 1162

# 2nd test execution
expected 1000, got 956
expected 1000, got 936
```

image

## Tasks
- [x] Support demo project
- [x] Confirm the API.md file contains a clear definition of what `spans` are required for the `foo` and `bar` entrypoints.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -98,6 +98,7 @@ dockerized-test:
 	  -e ES_URL=http://elasticsearch:9200 \
 	  -e KIBANA_URL=http://kibana:5601 \
 	  -e DJANGO_URL="http://djangoapp:8003" \
+	  -e DOTNET_URL="http://dotnetapp:80" \
```
Comment (v1v) [inline Makefile]: Let's review whether it's the internal or external port, if the second one then: https://github.com/elastic/apm-integration-testing/pull/424/files#diff-d0df5dd5f0019c3b878fde30f4b7341bR1356

File: Makefile [hunk 2]
```diff
@@ -98,6 +98,7 @@ dockerized-test:
 	  -e ES_URL=http://elasticsearch:9200 \
 	  -e KIBANA_URL=http://kibana:5601 \
 	  -e DJANGO_URL="http://djangoapp:8003" \
+	  -e DOTNET_URL="http://dotnetapp:8100" \
```
Comment (v1v) [inline Makefile]:
```
docker run  --name=apm-integration-testing1  --network=apm-integration-testing  --security-opt seccomp=unconfined  --rm  --entrypoint '' -ti apm-integration-testing /bin/bash
tester@daf2b7882782:~$ curl http://dotnetapp:80/
OKtester@daf2b7882782:~$ curl http://dotnetapp:8100
curl: (7) Failed to connect to dotnetapp port 8100: Connection refused
```

File: Makefile (hunk 1 above)
Comment (v1v) [reply Makefile]:
```
docker run  --name=apm-integration-testing1  --network=apm-integration-testing  --security-opt seccomp=unconfined  --rm  --entrypoint '' -ti apm-integration-testing /bin/bash
tester@daf2b7882782:~$ curl http://dotnetapp:80/
OKtester@daf2b7882782:~$ curl http://dotnetapp:8100
curl: (7) Failed to connect to dotnetapp port 8100: Connection refused
```

**Comment by gregkalapos** [comment]
Investigating failing tests here: https://github.com/elastic/apm-agent-dotnet/issues/215

**Pull request merged**